### PR TITLE
Consider initializer value of `NodeArg*` when comparing equivalence classes

### PR DESCRIPTION
**Pull request by sriduth**
**Description**: Adds a patch to compare initializers values when checking if 2 EquivalenceClasses are the same in the constant subexpression elimination optimization introduced in https://github.com/microsoft/onnxruntime/pull/3047, 

**Motivation and Context**
- Why is this change required? What problem does it solve?

Currently CSE is implemented such that all `NodeArgs*` to a Op node are considered distinct, consider the following pytorch
model - 

```python3
import torch
import torch.nn.functional as F
import numpy.random as npr

class TestCSE(torch.nn.Module):
  def __init__(self):
    super(TestCSE, self).__init__()
    self.outputs = []

  def forward(self, x):

    for i in range(0, 5):
      weights = torch.from_numpy(npr.rand(1,3,3,3))

      # even and odd convolutions will have same pad
      # so all even and odd convolution;s parent pad
      # must belong to the same EquivalenceClass
      padded_x  = F.pad(x, (i%2 + 1, i%2 + 1, i%2 + 1, i%2 + 1))
      out = F.conv2d(padded_x, weights)
      self.outputs.append(out)

    return self.outputs

torch.onnx.export(TestCSE(), torch.from_numpy(npr.rand(1, 3, 5, 5)), "ex2.onnx", verbose=True)
```

Generates an onnx model as follows:

Screenshot 2020-10-26 at 8 11 56 PM

Since every other `F.pad` operation pads by the same amount, they can be collected into 2 common groups and the resulting graph can be optimized to - 

Screenshot 2020-10-26 at 8 13 19 PM

A specific real use case is presented in #5585 

- If it fixes an open issue, please link to the issue here. - #5585

### CODE REVIEW

File: onnxruntime/core/optimizer/common_subexpression_elimination.cc [hunk 1]
```diff
@@ -365,12 +407,19 @@ Status CommonSubexpressionElimination::ApplyImpl(Graph& graph, bool& modified, i
     for (const NodeArg* input_def : node->InputDefs()) {
       auto it = equivalence_classes.find(input_def);
       if (it == equivalence_classes.end()) {
+        const onnx::TensorProto* initializer_ = new onnx::TensorProto();
```
Comment (yuslepukhin) [inline onnxruntime/core/optimizer/common_subexpression_elimination.cc:410]: >er_ [](start = 42, length = 3)

This is not a member

Comment (yuslepukhin) [inline onnxruntime/core/optimizer/common_subexpression_elimination.cc:410]: > = new onnx::TensorProto() [](start = 45, length = 26)

Who owns it?

Comment (sriduth) [reply onnxruntime/core/optimizer/common_subexpression_elimination.cc]: Hi @yuslepukhin - thanks for the review! - I have made the fix for this

File: onnxruntime/core/optimizer/common_subexpression_elimination.cc [hunk 2]
```diff
@@ -260,20 +269,52 @@ bool EquivalenceClass::operator==(const EquivalenceClass& other) const {
...
+      are_input_same = are_input_same && (iclasses_.size() == oclasses_.size());
+      
+      if(!are_input_same)
+        break;
+      
+      for(unsigned long j=0; j<iclasses_.size(); j++) {
+        const EquivalenceClass* clsi = iclasses_[j];
+        const EquivalenceClass* clso = oclasses_[j];
+        
+        if((clsi->initializer_ != nullptr) && (clso->initializer_ != nullptr)) {
```
Comment (sriduth) [inline onnxruntime/core/optimizer/common_subexpression_elimination.cc:288]: If we compare initializers in a pairwise manner we can end up omitting nodes that _can potentially be part of the same equivalence class_, say for example a elementwise Mul operation `Mul(x, 2) == Mul(2, x)` where x is some tensor. 

We should have some mechanism to handle nodes that follow commutative property - will check in a test case for the same. What do you think @yuslepukhin ?

Comment (sriduth) [reply onnxruntime/core/optimizer/common_subexpression_elimination.cc:288]: Test case:
```python3
import torch
import torch.nn.functional as F
import numpy.random as npr

class TestCSE(torch.nn.Module):
  def __init__(self):
    super(TestCSE, self).__init__()
    self.outputs = []

  def forward(self, x):
    mult_const = torch.from_numpy(npr.rand(1,3,5,5))
    for i in range(0, 5):
      weights = torch.from_numpy(npr.rand(1,3,3,3))

      if i%2 ==0:
        y = mult_const * x
      else:
        y = x * mult_const
      # even and odd convolutions will have same pad
      # so all even and odd convolution;s parent pad
      # must belong to the same EquivalenceClass
      padded_x  = F.pad(y, (i%2 + 1, i%2 + 1, i%2 + 1, i%2 + 1))
      out = F.conv2d(padded_x, weights)
      self.outputs.append(out)

    return self.outputs

torch.onnx.export(TestCSE(), torch.from_numpy(npr.rand(1, 3, 5, 5)), "ex3.onnx", verbose=True)
```
Input net - 

Screenshot 2020-10-28 at 12 53 10 PM

Output net - 

Screenshot 2020-10-28 at 12 53 33 PM

File: onnxruntime/core/optimizer/common_subexpression_elimination.cc [hunk 3]
```diff
@@ -132,7 +134,14 @@ class EquivalenceClass {
   // Currently, different inputs/initializers are always considered different values, although we
```
Comment (MaximKalininMS) [inline onnxruntime/core/optimizer/common_subexpression_elimination.cc:134]: Remove this comment?

File: onnxruntime/core/optimizer/common_subexpression_elimination.cc [hunk 4]
```diff
@@ -260,20 +269,52 @@ bool EquivalenceClass::operator==(const EquivalenceClass& other) const {
     return true;
   }
 
+  bool are_input_same = inputs_.size() == other.inputs_.size();
+  
+  if(are_input_same) {
+    for(unsigned long i=0; i<inputs_.size(); i++) {
```
Comment (MaximKalininMS) [inline onnxruntime/core/optimizer/common_subexpression_elimination.cc:275]: Note that currently an invariant is maintained that we never create two distinct but equal EquivalenceClasses. For operations, this is done by looking up a potentially new EquivalenceClass in `value_to_representative` map. You should do the same for initializers. Then this loop is not required.
Commutative operations can be handled by sorting the arguments in EquivalenceClass constructor (`Normalize` function should be a good place for that). Sorting by address is sufficient due to the above invariant.

File: onnxruntime/core/optimizer/common_subexpression_elimination.cc [hunk 5]
```diff
@@ -260,20 +269,52 @@ bool EquivalenceClass::operator==(const EquivalenceClass& other) const {
...
+        }
+      }
+    }
+  }
+
   // Below we compare inputs_ as pointers. This is valid due to how EquivalenceClass'es are constructed:
   // we'll never have two distinct but equal inputs_ here, so their addresses are effectively their value numbers.
   return hash_ == other.hash_ && output_index_ == other.output_index_ && discriminator_ == other.discriminator_ &&
-         non_op_value_ == other.non_op_value_ &&
+         //non_op_value_ == other.non_op_value_ &&
```
Comment (MaximKalininMS) [inline onnxruntime/core/optimizer/common_subexpression_elimination.cc:301]: We need to compare the initializers here.

File: onnxruntime/core/optimizer/common_subexpression_elimination.cc [hunk 6]
```diff
@@ -297,6 +338,7 @@ std::size_t EquivalenceClass::CalculateHash() const {
 // For inputs and constant initializers, output_index == kInvalidOutputIndex.
 struct Representative {
   const NodeArg* node_arg;
+  const onnx::TensorProto* initializer;
```
Comment (MaximKalininMS) [inline onnxruntime/core/optimizer/common_subexpression_elimination.cc:341]: I think this isn't used anywhere.

File: onnxruntime/core/optimizer/common_subexpression_elimination.cc [hunk 7]
```diff
@@ -386,13 +436,14 @@ Status CommonSubexpressionElimination::ApplyImpl(Graph& graph, bool& modified, i
...
       auto equivalence_class = onnxruntime::make_unique<EquivalenceClass>(*node, input_values, output_index, discriminator);
+      
       auto* raw_ptr = equivalence_class.get();
 
       auto it = value_to_representative.find(raw_ptr);
       if (it == value_to_representative.end()) {
         unique_equivalence_classes.push_back(std::move(equivalence_class));
         it = value_to_representative.emplace_hint(it, raw_ptr,
-                                                  Representative{output_def, node_index, output_index});
+                                                  Representative{output_def, nullptr, node_index, output_index});
```
Comment (MaximKalininMS) [inline onnxruntime/core/optimizer/common_subexpression_elimination.cc:446]: The assert below ` ORT_ENFORCE(representative.output_index != kInvalidOutputIndex);` will fire for duplicate initializers. It should be converted to an if condition, and ideally the duplicated initializer should be removed from the graph.

File: onnxruntime/core/optimizer/common_subexpression_elimination.cc [hunk 8]
```diff
@@ -260,20 +269,52 @@ bool EquivalenceClass::operator==(const EquivalenceClass& other) const {
...
 }
 
 std::size_t EquivalenceClass::CalculateHash() const {
   std::size_t hash = 0;
   UpdateHash(output_index_, hash);
   UpdateHash(discriminator_, hash);
-  UpdateHash(non_op_value_, hash);
+  
+  if(initializer_ != nullptr) {
+    UpdateHash(initializer_->raw_data(), hash);
```
Comment (MaximKalininMS) [inline onnxruntime/core/optimizer/common_subexpression_elimination.cc:313]: I think comparing just `raw_data` is insufficient. Check documentation for `TensorProto` [here](https://github.com/onnx/onnx/blob/master/onnx/onnx.proto), you need to compare at least the shape (`dims`), and probably `float_data`, `int32_data` etc...

Comment (MaximKalininMS) [reply onnxruntime/core/optimizer/common_subexpression_elimination.cc:313]: Also, I don't know if something special needs to be done in training. @yuslepukhin, are trained parameters represented as inputs (not initializers) during training?

File: onnxruntime/core/optimizer/common_subexpression_elimination.cc [hunk 9]
```diff
@@ -365,12 +407,20 @@ Status CommonSubexpressionElimination::ApplyImpl(Graph& graph, bool& modified, i
     for (const NodeArg* input_def : node->InputDefs()) {
       auto it = equivalence_classes.find(input_def);
       if (it == equivalence_classes.end()) {
+        const ONNX_NAMESPACE::TensorProto* initializer = nullptr;
+        
+        if(graph.IsInitializedTensor(input_def->Name())) {
```
Comment (MaximKalininMS) [inline onnxruntime/core/optimizer/common_subexpression_elimination.cc:412]: Need to also check that the initializer cannot be overridden ("initializers as inputs"). See `graph.GetOverridableInitializers()`.

File: onnxruntime/core/optimizer/common_subexpression_elimination.cc [hunk 10]
```diff
@@ -90,11 +90,12 @@ class EquivalenceClass {
 
   friend struct ::std::hash<EquivalenceClass>;
   friend std::vector<std::vector<const EquivalenceClass*>> Normalize(const Node& node, const std::vector<const EquivalenceClass*>& inputs);
-
-  explicit EquivalenceClass(const NodeArg* non_op_value)
+  
+  explicit EquivalenceClass(const NodeArg* non_op_value, const onnx::TensorProto* initializer_)
```
Comment (MaximKalininMS) [inline onnxruntime/core/optimizer/common_subexpression_elimination.cc:94]: Add a third constructor instead that accepts only the initializer. `non_op_value_` should then be renamed to `graph_input_` and set to nullptr in that constructor.

File: onnxruntime/core/optimizer/common_subexpression_elimination.cc (hunk 4 above)
Comment (sriduth) [reply onnxruntime/core/optimizer/common_subexpression_elimination.cc]: @MaximKalininMS, regarding the sorting of inputs - will it work for operations like `Div` if we always sort the inputs?

Comment (MaximKalininMS) [reply onnxruntime/core/optimizer/common_subexpression_elimination.cc]: Inputs should be sorted only for commutative operations (for which you want to merge `op(x, y)` and `op(y, x)`).

File: onnxruntime/core/optimizer/common_subexpression_elimination.cc [hunk 11]
```diff
@@ -412,7 +529,20 @@ Status CommonSubexpressionElimination::ApplyImpl(Graph& graph, bool& modified, i
          output_idx < end; ++output_idx) {
       const NodeArg* output_def = node->OutputDefs()[output_idx];
 
-      const EquivalenceClass* equivalence_class = equivalence_classes.at(output_def);
+      const EquivalenceClass* equivalence_class = nullptr;
+      
+      if(graph.IsInitializedTensor(output_def->Name())) {
```
Comment (MaximKalininMS) [inline onnxruntime/core/optimizer/common_subexpression_elimination.cc:534]: Can this ever happen? My understanding is that `output_def` here is an output of the `node`, so it cannot be an initializer or a graph input.

File: onnxruntime/core/optimizer/common_subexpression_elimination.cc [hunk 12]
```diff
@@ -350,7 +446,8 @@ Status CommonSubexpressionElimination::ApplyImpl(Graph& graph, bool& modified, i
   // This is the inverse of the above mapping, except that different NodeArgs can belong to the same
   // equivalence class. In that case these NodeArgs will be "merged" into one.
   std::unordered_map<const NodeArg*, const EquivalenceClass*, NodeArgPtrHash, NodeArgPtrEquality> equivalence_classes;
-
+  std::unordered_map<const ONNX_NAMESPACE::TensorProto*, const EquivalenceClass*, TensorProtoHash, TensorProtoEquality> eq;
```
Comment (MaximKalininMS) [inline onnxruntime/core/optimizer/common_subexpression_elimination.cc:449]: Why do we need a separate map? Can't we use `equivalence_classes` for everything?

File: onnxruntime/core/optimizer/common_subexpression_elimination.cc (hunk 11 above)
Comment (sriduth) [reply onnxruntime/core/optimizer/common_subexpression_elimination.cc:534]: I missed it completely - will try to wrap it in the next set of changes.

File: onnxruntime/core/optimizer/common_subexpression_elimination.cc [hunk 13]
```diff
@@ -90,30 +101,36 @@ class EquivalenceClass {
...
+        graph_input_(nullptr),
+        initializer_(nullptr),
         discriminator_(discriminator),
         hash_(CalculateHash()) {
   }
 
  private:
   std::size_t CalculateHash() const;
 
+  const std::string name_;
```
Comment (MaximKalininMS) [inline onnxruntime/core/optimizer/common_subexpression_elimination.cc:132]: If this is used only for debugging purposes, add a corresponding comment.

File: onnxruntime/core/optimizer/common_subexpression_elimination.cc [hunk 14]
```diff
@@ -170,6 +193,47 @@ bool AreRangesEqual(const Range& lhs, const Range& rhs) {
...
+    case ONNX_NAMESPACE::TensorProto_DataType_FLOAT:
+    case ONNX_NAMESPACE::TensorProto_DataType_COMPLEX64:
+      return AreRangesEqual(lhs->float_data(), rhs->float_data());
+      
+    case ONNX_NAMESPACE::TensorProto_DataType_DOUBLE:
+    case ONNX_NAMESPACE::TensorProto_DataType_COMPLEX128:
+      return AreRangesEqual(lhs->double_data(), rhs->double_data());
+      
+    default:
+      return AreRangesEqual(lhs->raw_data(), rhs->raw_data());
```
Comment (MaximKalininMS) [inline onnxruntime/core/optimizer/common_subexpression_elimination.cc:233]: I guess we should return false for safety here (in case another datatype and the corresponding `_data` field is added at some point).

File: onnxruntime/core/optimizer/common_subexpression_elimination.cc [hunk 15]
```diff
@@ -90,30 +101,36 @@ class EquivalenceClass {
 
   friend struct ::std::hash<EquivalenceClass>;
   friend std::vector<std::vector<const EquivalenceClass*>> Normalize(const Node& node, const std::vector<const EquivalenceClass*>& inputs);
-
-  explicit EquivalenceClass(const NodeArg* non_op_value)
-      : attributes_(nullptr),
+    
+  explicit EquivalenceClass(std::shared_ptr<struct NodeArgMaybeValue> non_op_value)
```
Comment (MaximKalininMS) [inline onnxruntime/core/optimizer/common_subexpression_elimination.cc:105]: Nit: `struct` is unnecessary here and in several other places below.

File: onnxruntime/core/optimizer/common_subexpression_elimination.cc [hunk 16]
```diff
@@ -170,6 +193,47 @@ bool AreRangesEqual(const Range& lhs, const Range& rhs) {
          std::equal(lhs.begin(), lhs.end(), rhs.begin());
 }
 
+
+bool AreEqual(const ONNX_NAMESPACE::TensorProto* lhs, const ONNX_NAMESPACE::TensorProto* rhs)
+{
+  if((lhs == nullptr) || (rhs == nullptr))
```
Comment (MaximKalininMS) [inline onnxruntime/core/optimizer/common_subexpression_elimination.cc:199]: Could start with: `if (lhs == rhs) return true`.

File: onnxruntime/core/optimizer/common_subexpression_elimination.cc [hunk 17]
```diff
@@ -156,7 +179,7 @@ std::vector<std::vector<const EquivalenceClass*>> Normalize(const Node& node, co
     }
 
     // Remove missing optional inputs from the back
-    while (!arg.empty() && arg.back()->output_index_ == kInvalidOutputIndex && arg.back()->non_op_value_ == nullptr) {
+    while (!arg.empty() && arg.back()->output_index_ == kInvalidOutputIndex && ((arg.back()->graph_input_ == nullptr) && (arg.back()->initializer_ == nullptr))) {
```
Comment (MaximKalininMS) [inline onnxruntime/core/optimizer/common_subexpression_elimination.cc:182]: Nit: unnecessary parenthesis.

**Review by MaximKalininMS — COMMENTED** [review]
It seems that there are unnecessary submodule changes (probably side effect of a merge).

**Comment by snnn** [comment]
Close inactive PRs with merge conflicts
